ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/HiveMetaStoreAuthorizer.java

```diff
@@ -0,0 +1,234 @@
+      @Override
+      protected HiveMetastoreAuthenticationProvider initialValue() {
+    try {
+      return (HiveMetastoreAuthenticationProvider) HiveUtils.getAuthenticator(tConfig.get(), HiveConf.ConfVars.HIVE_METASTORE_AUTHENTICATOR_MANAGER);
+    } catch (HiveException excp) {
+      throw new IllegalStateException("Authentication provider instantiation failure", excp);
+    }
+      }
+  };
+
+  public static synchronized  PreEventContext getPreEventContext() {
+      return preEventContext;
+  }
+  public HiveMetaStoreAuthorizer(Configuration config) {
+    super(config);
+    superUsers = config.getStrings(HIVE_SUPER_USERS);
+  }
+
+  @Override
+  public final void onEvent(PreEventContext preEventContext) throws MetaException, NoSuchObjectException, InvalidOperationException {
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("==> HiveMetaStoreAuthorizer.onEvent(): EventType=" + preEventContext.getEventType());
+    }
+
+    HiveMetaStoreAuthzInfo authzContext = buildAuthzContext(preEventContext);
+
+    if (!skipAuthorization(authzContext)) {
+      try {
+        HiveConf hiveConf          = new HiveConf(super.getConf(), HiveConf.class);
+        HiveAuthorizerFactory authorizerFactory = HiveUtils.getAuthorizerFactory(hiveConf, HiveConf.ConfVars.HIVE_AUTHORIZATION_MANAGER);
+
+        if (authorizerFactory != null) {
+          HiveMetastoreAuthenticationProvider authenticator = tAuthenticator.get();
+
+          authenticator.setConf(hiveConf);
+
+          HiveAuthzSessionContext.Builder authzContextBuilder = new HiveAuthzSessionContext.Builder();
+
+          authzContextBuilder.setClientType(HiveAuthzSessionContext.CLIENT_TYPE.HIVECLI);
+          authzContextBuilder.setSessionString("HiveMetaStore");
+
+          HiveAuthzSessionContext authzSessionContext = authzContextBuilder.build();
+
+          HiveAuthorizer hiveAuthorizer = authorizerFactory.createHiveAuthorizer(new HiveMetastoreClientFactoryImpl(), hiveConf, authenticator, authzSessionContext);
+
+          authenticator.getUserName();
+
+          checkPrivileges(authzContext, hiveAuthorizer);
+        }
+      } catch (HiveException e) {
+        LOG.error("HiveMetaStoreAuthorizer.onEvent(): failed", e);
+      }
+    }
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("<== HiveMetaStoreAuthorizer.onEvent(): EventType=" + preEventContext.getEventType());
+    }
+  }
+
+
+  private HiveMetaStoreAuthzInfo buildAuthzContext(PreEventContext preEventContext) throws MetaException {
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("==> HiveMetaStoreAuthorizer.buildAuthzContext(): EventType=" + preEventContext.getEventType());
+    }
+
+    HiveMetaStoreAuthorizableEvent authzEvent = null;
+
+    if (preEventContext != null) {
+      this.preEventContext = preEventContext;
+      switch (preEventContext.getEventType()) {
+        case CREATE_DATABASE:
```

Comment:
There are missing events need handling, some of those also require invoking preListener in HiveMetaStore.java
1. create catalog, alter catalog, drop catalog*, admin only
2. drop constrain*, add primary key/foreign key/unique key/not null/default/check constraint*, these require alter table privs
3. exchange partition*, require alter partition privs
4. update/delete table column stats*, require alter table privs
5. update/delete partition column stats*, require alter partition privs
6. create/drop/alter function*, require function privs (if there's any)
7. schema registry related event: add/alter/drop schema, add serde, maybe we can skip for now, but need to track for the future
8. all auth related events: PreAuthorizationCallEvent

Reply:
Comment 7 -> add/alter/drop schema -> All these commands will not be authorized for the users 

Comment 8 -> PreAuthorizationCallEvent -> All PreAuthorizationCallEvent call will not be authorized for the users.

Comment  2 to 6 -> all these needs changes to HiveMetastore to fire the auth event and followed authorization changes. I shall address this in subsequent JIRA once this patch is completed and validated. This needs a signification change and verification for it.

Comment 1 -> CREATE CATALOG / ALTER CATALOG / DROP CATALOG can be done as admin user which is already in place. Rest all users will not be authorized to do these operations.

CREATE VIEW / ALTER VIEW / DROP VIEW - operations related to views will not authorized for the users. Only admin users are allowed to do.